{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Создаём папку для выходных файлов. Если таковая уже имеется, то очищаем ее от старых файлов.\n",
    "if os.path.exists(\"Output\"):\n",
    "    for file in os.listdir(\"Output/\"):\n",
    "        os.remove(\"Output/\" + file)    \n",
    "else:        \n",
    "    os.mkdir(\"Output\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cчитаем количество строк, отведённых под тэги [header section]\n",
    "def Skip_string (infile):\n",
    "    count = 0\n",
    "    with open (infile) as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"@\"):\n",
    "                count = count + 1\n",
    "            else:\n",
    "                break\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Функция, получающая таблицу формата Bedgraph из листа, содержащего информацию о покрытии каждого элемента\n",
    "# chr 10, [1 1 1 1 3 3 0 0 4 4 4] (здесь первый элемент имеет нулевой индекс)\n",
    "#          _______ \n",
    "#              |            \n",
    "#  >> [10 0 4 4]   ___\n",
    "#                   |\n",
    "#          [10 4 6 2]      ____\n",
    "#                           |\n",
    "#                  [10 8 11 3]\n",
    "\n",
    "def Make_bedgraph(cover_list):\n",
    "    \n",
    "    flg = 0 # необходим для пропуска баз с нулевым покрытием, эти регионы не учитываются в bedgraph файле\n",
    "    \n",
    "    # отдельно считаем покрытие для первой базы из исходного файла \n",
    "    if cover_list[0]!= 0:\n",
    "        bed_list[0].append(chrom)\n",
    "        bed_list[1].append(0)\n",
    "        bed_list[3].append(cover_list[0])\n",
    "        flg = 1\n",
    "        \n",
    "    for i in range(1, len(cover_list)):\n",
    "        #print(i)\n",
    "        if cover_list[i] != cover_list[i-1]: \n",
    "            if flg!=0:\n",
    "                bed_list[2].append(i)  # добавляем конец предыдущего полуинтервала покрытия = началу нового отрезка покрытия. \n",
    "            if cover_list[i]!= 0:  \n",
    "                bed_list[0].append(chrom)\n",
    "                bed_list[1].append(i)\n",
    "                bed_list[3].append(np.int(cover_list[i]))\n",
    "                flg = 1\n",
    "            if cover_list[i] == 0:\n",
    "                flg = 0\n",
    "                \n",
    "    return(bed_list)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File:  test.sam\n",
      "Chromosomes:  [ 1  2 10]\n",
      "Calculating chromosome 1...\n",
      "Calculating chromosome 2...\n",
      "Calculating chromosome 10...\n",
      "    chr     start       end  value\n",
      "0     1   1919128   1919172      1\n",
      "1     1   1989744   1989774      1\n",
      "2     1   6461211   6461242      1\n",
      "3     1   7973788   7973818      1\n",
      "4     1  18719173  18719247      1\n",
      "5     1  20586084  20586121      1\n",
      "6     1  21716735  21716784      1\n",
      "7     1  28226012  28226050      1\n",
      "8     1  33119239  33119276      1\n",
      "9     1  33669900  33669940      1\n",
      "10    1  35614051  35614134      1\n",
      "11    1  37140074  37140104      1\n",
      "12    1  39367850  39367892      1\n",
      "13    1  40023765  40023804      1\n",
      "14    1  40585941  40585974      1\n",
      "15    1  44166581  44166612      1\n",
      "16    1  44581128  44581167      1\n",
      "17    1  44927572  44927616      1\n",
      "18    1  45305040  45305071      1\n",
      "19    1  46757215  46757247      1\n",
      "20    1  48619046  48619183      1\n",
      "21    1  48619598  48619749      1\n",
      "22    1  48620380  48620497      1\n",
      "23    1  48620497  48620531      2\n",
      "24    1  48620531  48620648      1\n",
      "25    2   3637657   3637692      1\n",
      "26    2   5338065   5338216      1\n",
      "27    2   5338308   5338459      1\n",
      "28    2   6421418   6421451      1\n",
      "29    2  12662980  12663131      1\n",
      "30    2  12663206  12663344      1\n",
      "31    2  14007802  14007889      1\n",
      "32    2  14007889  14007940      2\n",
      "33    2  14007940  14008040      1\n",
      "34    2  16731255  16731285      1\n",
      "35    2  21290556  21290557      1\n",
      "36    2  21290557  21290603      3\n",
      "37    2  21290603  21290604      2\n",
      "38    2  21290604  21290607      1\n",
      "39    2  27057236  27057270      1\n"
     ]
    }
   ],
   "source": [
    "in_dir = \"Input/\"\n",
    "read_len = 300 # Default = 300\n",
    "\n",
    "for file in os.listdir(in_dir): # Ведём рассчеты для всех файлов, помещенных в папку Input\n",
    "    print(\"File: \", file)\n",
    "    bed_list = [[]*50 for x in range(4)]\n",
    "    tags_num = Skip_string(in_dir + file) \n",
    "    bam = pd.read_csv(in_dir + file, delimiter = \"\\t\", usecols = [2, 3, 5], skiprows = tags_num, names = [\"chr\", \"coord\", \"CIGAR\"])\n",
    "    bam = bam.sort_values(by = [\"chr\", \"coord\"]) # cортируем данные сначала по хромосомам, затем - в пределах каждой из хромосом\n",
    "    bam = bam.reset_index(drop = True) # заново нумеруем таблицу\n",
    "\n",
    "    \n",
    "    chroms = bam.chr.unique() # получаем список хромосом из .sam файла\n",
    "    print(\"Chromosomes: \", chroms)\n",
    "\n",
    "    \n",
    "    for chrom in chroms:\n",
    "        \n",
    "        print(\"Calculating chromosome {}...\".format(chrom))\n",
    "        bam_for_chr = bam[bam['chr']== chrom]['coord']\n",
    "        bam_chr_indices = bam_for_chr.index.tolist()\n",
    "        #print(bam_chr_indices)\n",
    "        \n",
    "        \n",
    "        coverage_all = np.zeros((max(bam_for_chr) + read_len)) # cоздаем лист, который затем заполним информацией о покрытии каждого элемента\n",
    "\n",
    "        \n",
    "#         M alignment match (can be a sequence match or mismatch)\n",
    "#         I insertion to the reference\n",
    "#         D deletion from the reference\n",
    "#         N skipped region from the reference\n",
    "#         S soft clipping (clipped sequences present in SEQ)\n",
    "#         H hard clipping (clipped sequences NOT present in SEQ)\n",
    "#         P padding (silent deletion from padded reference)\n",
    "\n",
    "    \n",
    "        for i in bam_chr_indices:\n",
    "            \n",
    "            # Распознавание CIGAR операторов\n",
    "            cigar = re.findall(r'(\\d+)([A-Z]{1})', bam.CIGAR[i])   #cigar - лист вида: [('13', 'S'), ('138', 'M')]  \n",
    "            # print(cigar)\n",
    "            cursor = bam.coord[i]\n",
    "            \n",
    "            for c in cigar:\n",
    "                if c[1] == \"M\" or c[1] == \"=\" or c[1] == \"X\":  # обрабатываем части ридов, которые заматчились\n",
    "                    coord_start = cursor\n",
    "                    coord_end = coord_start + int(c[0])\n",
    "                    cursor = coord_end\n",
    "                    for i in range(coord_start-1, coord_end-1):\n",
    "                         coverage_all[i] += 1  \n",
    "                elif c[1] == \"D\" or c[1] == \"N\" or c[1] == \"P\":\n",
    "                    cursor = cursor + int(c[0])                \n",
    "\n",
    "        bed_list = Make_bedgraph(coverage_all)\n",
    "        \n",
    "    bed_df = pd.DataFrame(list(zip(*bed_list)),  columns = [\"chr\", \"start\", \"end\", \"value\"]) \n",
    "    print(bed_df[:40])\n",
    "\n",
    "    # Cохраняем файл \n",
    "    file = file.split(\".\")[0]\n",
    "    bed_df.to_csv(\"Output/{}.bedgraph\".format(file), sep = \"\\t\", index = False, header = False)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
