{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Создаём папку для выходных файлов. Если таковая уже имеется, то очищаем ее от старых файлов.\n",
    "if os.path.exists(\"Output\"):\n",
    "    for file in os.listdir(\"Output/\"):\n",
    "        os.remove(\"Output/\" + file)    \n",
    "else:        \n",
    "    os.mkdir(\"Output\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cчитаем количество строк, отведённых под тэги [header section]\n",
    "def Skip_string (infile):\n",
    "    count = 0\n",
    "    with open (infile) as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"@\"):\n",
    "                count = count + 1\n",
    "            else:\n",
    "                break\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Функция, получающая таблицу формата Bedgraph из листа, содержащего информацию о покрытии каждого элемента\n",
    "# chr 10, [1 1 1 1 3 3 0 0 4 4 4] (здесь первый элемент имеет нулевой индекс)\n",
    "#          _______ \n",
    "#              |            \n",
    "#  >> [10 0 4 4]   ___\n",
    "#                   |\n",
    "#          [10 4 6 2]      ____\n",
    "#                           |\n",
    "#                  [10 8 11 3]\n",
    "\n",
    "def Make_bedgraph(cover_list):\n",
    "    \n",
    "    flg = 0 # необходим для пропуска баз с нулевым покрытием, эти регионы не учитываются в bedgraph файле\n",
    "    \n",
    "    # отдельно считаем покрытие для первой базы из исходного файла \n",
    "    if cover_list[0]!= 0:\n",
    "        bed_list[0].append(chrom)\n",
    "        bed_list[1].append(0)\n",
    "        bed_list[3].append(cover_list[0])\n",
    "        flg = 1\n",
    "        \n",
    "    for i in range(1, len(cover_list)):\n",
    "        #print(i)\n",
    "        if cover_list[i] != cover_list[i-1]: \n",
    "            if flg!=0:\n",
    "                bed_list[2].append(i)  # добавляем конец предыдущего полуинтервала покрытия = началу нового отрезка покрытия. \n",
    "            if cover_list[i]!= 0:  \n",
    "                bed_list[0].append(chrom)\n",
    "                bed_list[1].append(i)\n",
    "                bed_list[3].append(np.int(cover_list[i]))\n",
    "                flg = 1\n",
    "            if cover_list[i] == 0:\n",
    "                flg = 0\n",
    "                \n",
    "    return(bed_list)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File:  for_test.sam\n",
      "Chromosomes:  ['III' 'IV' 'V']\n",
      "Calculating chromosome III...\n",
      "Calculating chromosome IV...\n",
      "Calculating chromosome V...\n",
      "    chr  start    end  value\n",
      "0   III   2555   2655      1\n",
      "1   III   3100   3200      1\n",
      "2   III   6790   6890      1\n",
      "3   III   7234   7334      1\n",
      "4   III  21931  22031      1\n",
      "5   III  22351  22451      1\n",
      "6   III  25570  25670      1\n",
      "7   III  25996  26096      1\n",
      "8   III  27361  27461      1\n",
      "9   III  27756  27856      1\n",
      "10  III  36606  36706      1\n",
      "11  III  36967  37067      1\n",
      "12  III  48187  48287      1\n",
      "13  III  48499  48599      1\n",
      "14  III  50455  50555      1\n",
      "15  III  50788  50888      1\n",
      "16  III  52456  52556      1\n",
      "17  III  52857  52957      1\n",
      "18  III  57527  57627      1\n",
      "19  III  57959  58059      1\n",
      "20  III  60423  60523      1\n",
      "21  III  60838  60938      1\n",
      "22  III  61596  61696      1\n",
      "23  III  61894  61994      1\n",
      "24  III  65328  65428      1\n",
      "25  III  65767  65867      1\n",
      "26  III  72590  72690      1\n",
      "27  III  73047  73147      1\n",
      "28  III  76940  77040      1\n",
      "29  III  77314  77414      1\n",
      "30  III  85056  85156      1\n",
      "31  III  85421  85521      1\n",
      "32  III  88832  88932      1\n",
      "33  III  89272  89372      1\n",
      "34  III  89702  89802      1\n",
      "35  III  90154  90254      1\n",
      "36  III  94128  94228      1\n",
      "37  III  94531  94631      1\n",
      "38  III  96342  96442      1\n",
      "39  III  96689  96789      1\n"
     ]
    }
   ],
   "source": [
    "in_dir = \"Input/\"\n",
    "read_len = 300 # Default = 300\n",
    "\n",
    "for file in os.listdir(in_dir): # Ведём рассчеты для всех файлов, помещенных в папку Input\n",
    "    print(\"File: \", file)\n",
    "    bed_list = [[]*50 for x in range(4)]\n",
    "    tags_num = Skip_string(in_dir + file) \n",
    "    bam = pd.read_csv(in_dir + file, delimiter = \"\\t\", usecols = [2, 3, 5], skiprows = tags_num, names = [\"chr\", \"coord\", \"CIGAR\"])\n",
    "    bam = bam.sort_values(by = [\"chr\", \"coord\"]) # cортируем данные сначала по хромосомам, затем - в пределах каждой из хромосом\n",
    "    bam = bam.reset_index(drop = True) # заново нумеруем таблицу\n",
    "\n",
    "    \n",
    "    chroms = bam.chr.unique() # получаем список хромосом из .sam файла\n",
    "    print(\"Chromosomes: \", chroms)\n",
    "\n",
    "    \n",
    "    for chrom in chroms:\n",
    "        \n",
    "        print(\"Calculating chromosome {}...\".format(chrom))\n",
    "        bam_for_chr = bam[bam['chr']== chrom]['coord']\n",
    "        bam_chr_indices = bam_for_chr.index.tolist()\n",
    "        #print(bam_chr_indices)\n",
    "        \n",
    "        \n",
    "        coverage_all = np.zeros((max(bam_for_chr) + read_len)) # cоздаем лист, который затем заполним информацией о покрытии каждого элемента\n",
    "\n",
    "        \n",
    "#         M alignment match (can be a sequence match or mismatch)\n",
    "#         I insertion to the reference\n",
    "#         D deletion from the reference\n",
    "#         N skipped region from the reference\n",
    "#         S soft clipping (clipped sequences present in SEQ)\n",
    "#         H hard clipping (clipped sequences NOT present in SEQ)\n",
    "#         P padding (silent deletion from padded reference)\n",
    "\n",
    "    \n",
    "        for i in bam_chr_indices:\n",
    "            \n",
    "            # Распознавание CIGAR операторов\n",
    "            cigar = re.findall(r'(\\d+)([A-Z]{1})', bam.CIGAR[i])   #cigar - лист вида: [('13', 'S'), ('138', 'M')]  \n",
    "            # print(cigar)\n",
    "            cursor = bam.coord[i]\n",
    "            \n",
    "            for c in cigar:\n",
    "                if c[1] == \"M\" or c[1] == \"=\" or c[1] == \"X\":  # обрабатываем части ридов, которые заматчились\n",
    "                    coord_start = cursor\n",
    "                    coord_end = coord_start + int(c[0])\n",
    "                    cursor = coord_end\n",
    "                    for i in range(coord_start-1, coord_end-1):\n",
    "                         coverage_all[i] += 1  \n",
    "                elif c[1] == \"D\" or c[1] == \"N\" or c[1] == \"P\":\n",
    "                    cursor = cursor + int(c[0])                \n",
    "\n",
    "        bed_list = Make_bedgraph(coverage_all)\n",
    "        \n",
    "    bed_df = pd.DataFrame(list(zip(*bed_list)),  columns = [\"chr\", \"start\", \"end\", \"value\"]) \n",
    "    print(bed_df[:40])\n",
    "\n",
    "    # Cохраняем файл \n",
    "    file = file.split(\".\")[0]\n",
    "    bed_df.to_csv(\"Output/{}.bedgraph\".format(file), sep = \"\\t\", index = False, header = False)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
